src/main/java/ua/olena/tracing/graph/Graph.java

```diff
@@ -0,0 +1,244 @@
+
+            Edge edge = edges.get(from + to);
+            if (edge != null) {
+                List<Integer> latencies = edge.getLatencies();
+                int average = calculateAverage(latencies);
+                latency = latency + average;
+            } else {
+                return ERROR_MSG;
+            }
+
+            idx = idx + 2;
+        }
+
+        return String.valueOf(latency);
+    }
+
+    /**
+     * Calculates how many traces there are between two services with maximum/exact number of stops.
+     *
+     * @param from Node to start from
+     * @param to Node to finish on
+     * @param stops Number of stops like 2 For A-B-C
+     * @param condition Maximum or Exact
+     * @return number of traces
+     */
+    public Integer getTracesByStops(String from, String to, Integer stops, Condition condition) {
+        int traceNumber = 0;
+        int depth = 0;
+
+        Stack<String> visited = new Stack<>();
+        Stack<String> stack = new Stack<>();
+        Stack<Integer> depthStack = new Stack<>();
+        stack.push(from);
+        depthStack.push(depth);
+        while (!stack.isEmpty()) {
+            String node = stack.pop();
+            depth = depthStack.pop() + 1;
+            visited.add(node);
+            Vertex vertex = vertexes.get(node);
+            Map<String, Vertex> neighbors = vertex.getNeighbors();
+
+            if (depth != stops + 1) {
+                pushNeighborsToStack(depth, stack, depthStack, neighbors);
+            } else {
+                goBack(visited, depthStack);
+            }
+
+            if (isReachedDestination(node, to, depth, stops, condition))
+                traceNumber++;
+        }
+
+        return traceNumber;
+    }
+
+    /**
+     * Get shortest trace between two nodes for provided graph.
+     *
+     * @param from Node to start from
+     * @param to Node to finish at
+     * @return Minimal latency of the path between two nodes
+     */
+    public Integer getShortestTrace(String from, String to) {
+        int minLatency = 0;
+        int depth = 0;
+
+        Stack<String> visited = new Stack<>();
+        Stack<String> stack = new Stack<>();
+        Stack<Integer> depthStack = new Stack<>();
+        stack.push(from);
+        depthStack.push(depth);
+        while (!stack.isEmpty()) {
+            String node = stack.pop();
+            depth = depthStack.pop() + 1;
+            visited.add(node);
+            Vertex vertex = vertexes.get(node);
+            Map<String, Vertex> neighbors = vertex.getNeighbors();
+
+            //If I'm going from C to C and I'm in C but this is starting point and not destination
+            if (!node.equals(to) || visited.size() == 1) {
+                for (String neighbor : neighbors.keySet()) {
+                    if (!visited.contains(neighbor) || (visited.contains(neighbor) && neighbor.equals(to))) {
+                        stack.push(neighbor);
```

Comment:
By pushing the neighbours on the stack you're creating the next iteration loop.
So if you're on "A" you have neighbours "B" and "E". Then you push  "B" and "E" on the stack to check later (unless they're in the "visited" stack.

So fine so good. 

Tree is:
```
AB5, BC4, AD5, DC8, DE6 .....

 [A] ---5---> [B] ---4---> [C] 
    \---5---> [D] ---8---> [C]
               \---6--->[E]
```

So you figure out after a while that the path A-B-C has latency "9". But you're not using this information for your next iteration steps.

You could optimise the code in that way that you evaluate the latency before pushing a vertex to the stack.
So suppose you travelled back to A-D. In this case latency between A and D is 5. 

If you now would do a look ahead for neighbours' latency then you know that next neighbour D-C has 8 and D-E has 6... So you already know A-D-C has latency 13 and A-D-E has latency 11. And both are higher than 9, so you don't need to follow this track anymore. For huge graphs this might save you some extra hops.

Reply:
Cache! Cache? I think it's cache I need here. 
I'm not sure how to easily push A-D-C calculations to cache but I can make it easier by pushing average traces to a cache (especially taken that there are not that many services in the example, however I'll have to invalidate it on adding a new edge...) Thanks!

Reply:
Added cache to at least not recalculate the same thing over and over. And invalidation after adding a new edge (as it's a map, no issues I expect there)